utils/data-migration/migrate-acbs-utilisation-data-to-tfm/index.ts

```diff
@@ -0,0 +1,93 @@
+import { existsSync, readFileSync } from 'fs';
+import { resolve } from 'path';
+
+/**
+ * The file path to the JSON file containing ACBS utilisation data.
+ */
+const JSON_FILE_PATH = resolve(__dirname, 'acbs-utilisation-data.json');
+
+/**
+ * The report period for the utilisation data.
+ */
+const REPORT_PERIOD: ReportPeriodPartialEntity = {
+  start: { month: 1, year: 2020 },
+  end: { month: 3, year: 2020 },
+};
+
+type UtilisationDataEntry = {
+  facilityId: string;
+  utilisation: number;
+  fixedFeePremium: number;
+};
+
+/**
+ * Validates the report period to ensure the end date is not before the start date.
+ * @param reportPeriod - The report period to validate.
+ * @throws {Error} If the end date is before the start date.
+ */
+const validateReportPeriod = (reportPeriod: ReportPeriodPartialEntity): void => {
+  const { start, end } = reportPeriod;
+
+  const isStartYearAfterEndYear = start.year > end.year;
+  const isStartMonthAfterEndMonthInSameYear = start.year === end.year && start.month > end.month;
+
+  if (isStartYearAfterEndYear || isStartMonthAfterEndMonthInSameYear) {
+    throw new Error('Report period end date cannot be before start date');
+  }
+};
+
+/**
+ * Inserts a facility utilisation entry into the database.
+ * @param dataSource - The data source for database operations.
+ * @param utilisationDataEntry - The utilisation data entry to insert.
+ * @param utilisationDataEntry.facilityId - The ID of the facility.
+ * @param utilisationDataEntry.utilisation - The utilisation value.
+ * @param utilisationDataEntry.fixedFeePremium - The fixed fee premium value.
+ */
+const insertFacilityUtilisationEntry = async (dataSource: DataSource, { facilityId, utilisation, fixedFeePremium }: UtilisationDataEntry) => {
+  const facilityUtilisationData = FacilityUtilisationDataEntity.create({
+    id: facilityId,
+    reportPeriod: REPORT_PERIOD,
+    utilisation,
+    fixedFee: fixedFeePremium,
+    requestSource: { platform: 'SYSTEM' },
+  });
+
+  await dataSource.manager.save(FacilityUtilisationDataEntity, facilityUtilisationData);
+};
+
+/**
+ * Runs the migration process to transfer ACBS utilisation data to TFM.
+ * @throws {Error} If the JSON input file is not found.
+ */
+const runMigrationOfAcbsUtilisationDataToTfm = async () => {
+  validateReportPeriod(REPORT_PERIOD);
+
+  const dataSource = await SqlDbDataSource.initialize();
+
+  try {
+    if (!existsSync(JSON_FILE_PATH)) {
+      throw new Error(`Migration JSON file not found at path: ${JSON_FILE_PATH}`);
+    }
+    const jsonData = readFileSync(JSON_FILE_PATH, 'utf8');
+
+    const utilisationData = JSON.parse(jsonData) as UtilisationDataEntry[];
+
+    await Promise.all(utilisationData.map((entry) => insertFacilityUtilisationEntry(dataSource, entry)));
+  } catch (error) {
+    console.error('Failed to migrate ACBS utilisation data to TFM: %o', error);
+  } finally {
+    await dataSource.destroy();
+  }
+};
+
+runMigrationOfAcbsUtilisationDataToTfm()
```

Comment:
This is much better adhering to promise pattern. Thank you.

Reply:
Just out of curiosity, what `Promise` pattern is this adhering to? As far as I know, it is not considered best practice to mix and match `.then` and `.catch` with `async`/`await` syntax due to the unexpected ways in which they can interact with each other, but maybe I am missing something?

Reply:
DTFS code base widely uses two patterns generally:
* Middleware
* Promise

Promise pattern is a widely used for async calls as they are design to handle returned promise resolves or rejects. as it keeps the code simple (KISS), DRY, adheres to SOLID principles and most importantly helps de-couple crucial logics and exceptional handling.

There are other patterns which provide similar advantages but provided wide use of this pattern in DTFS code base, it makes sense to use it.

Reply:
This is a script so I don't mind how this is written in this particular case but just wanted to check the above - in our codebase we almost entirely use async/await. E.g. in api.js files it's all async/await and then in our controllers (both old and new ones) we use async/await with async code not promises. 
Are there different cases where we are trying to use promises over the normal async/await we use in the core of our controller and API logic?